src/middleware/authentication.ts

```diff
@@ -0,0 +1,22 @@
+import { AuthenticationError } from "apollo-server";
+import { logger } from "../logger";
+import * as dotenv from "dotenv";
+
+dotenv.config();
+
+export const authenticationContextMiddleware = ({ req }: { req: any }) => {
+  const authHeaderUser: string = req.headers.authorization || "";
+  const expectedHeader: string = process.env.AUTH_TOKEN!;
+
+  if (authHeaderUser !== expectedHeader) {
+    logger.error(`Unauthorized access attempt: ${req.method} ${req.url}`);
+    throw new AuthenticationError("Unauthorized");
+  }
+  logger.info(
+    `Received request: ${req.method} ${req.url} with headers: ${JSON.stringify(
```

Comment:
Usually logs cost money and we tend to avoid logging every request. Instead, we generate arbitrary ids which helps us trace requests down the line. 

Another thing is that, right now you have a coupled middleware: Authentication + Logging. 
It's a good practice to have a separate middleware for handling the request logging or the creation of the above mentioned arbitrary id and appending it to the context.

Reply:
I decoupled them. In this application I didn't use express application and only apollo server, therefore those middlewares were combined in a function in app.ts files and passed as context. If you think that it is better idea to use express app and its middleware functionality I can do it.

Reply:
I used more or less the same approach in my code:

```
const createContext = ({ req }: {req : Request}) => {
  // Initialize an empty context object
  let context = {};

  // Apply authentication and logging middleware
  context = { ...context, ...authenticationContextMiddleware({ req }) };
  context = { ...context, ...loggingContextMiddleware({ req }) };

  return context;
};
```

Reply:
I might be late to the party, but yeah as far as I see, the apollo-server lib does use Express when creating the server